File: Two_pointers/Java/Sentence Similarity III.java
// Problem link: https://leetcode.com/problems/sentence-similarity-iii/description
// Time: O(n + m)
// Space: O(n + m)
class Solution {
    public boolean areSentencesSimilar(String sentence1, String sentence2) {
        String[] storeOne = sentence1.split(" ");
        String[] storeTwo = sentence2.split(" ");

        if(storeOne.length > storeTwo.length) {
            String[] temp = storeOne;
            storeOne = storeTwo;
            storeTwo = temp;
        }

        int left = 0, rightOne = storeOne.length - 1, rightTwo = storeTwo.length - 1;

        while(left < storeOne.length && storeOne[left].equals(storeTwo[left])) {
            left++;
        }

        while(rightOne >= left && storeOne[rightOne].equals(storeTwo[rightTwo])) {
            rightOne--;
            rightTwo--;
        }

        return left > rightOne;
    }
}

File: Two_pointers/Python/Sentence Similarity III.py
# Problem link: https://leetcode.com/problems/sentence-similarity-iii/description
# Time: O(n + m)
# Space: O(n)
class Solution:
    def areSentencesSimilar(self, sentence1: str, sentence2: str) -> bool:
        storeOne= sentence1.split()
        storeTwo = sentence2.split()

        if len(storeOne) > len(storeTwo):
            storeOne, storeTwo = storeTwo, storeOne

        left, rightOne, rightTwo = 0, len(storeOne) - 1, len(storeTwo) - 1

        while left < len(storeOne) and storeOne[left] == storeTwo[left]:
            left += 1

        while rightOne >= left and storeOne[rightOne] == storeTwo[rightTwo]:
            rightOne, rightTwo = rightOne - 1, rightTwo - 1
        
        return left > rightOne
